src/client/cl_entity.c

```diff
@@ -401,6 +392,39 @@ void Cl_Interpolate(void) {
 			ent->animation2.animation = ent->current.animation2 & ~ANIM_TOGGLE_BIT;
 			ent->animation2.time = cl.unclamped_time;
 		}
+
+		if (ent->current.solid > SOLID_DEAD) {
+			const vec_t *angles;
+
+			// FIXME
+			//
+			// Currently, the entity's collision state is snapped to the most recently received
+			// server frame, rather than its interpolated state. This works well with the prediction
+			// code, but has certain side effects, such as shadows flickering while riding platforms
+			// etc. Ideally, we would create the clipping matrix from ent->origin and ent->angles,
+			// but this introduces prediction error issues. Need to understand why the prediction
+			// code doesn't play well with the more accurate collision simulation.
+
+			if (ent->current.solid == SOLID_BSP) {
+				angles = ent->current.angles;
+
+				const r_model_t *mod = cl.model_precache[ent->current.model1];
+
+				assert(mod);
+				assert(mod->bsp_inline);
+
+				VectorCopy(mod->bsp_inline->mins, ent->mins);
+				VectorCopy(mod->bsp_inline->maxs, ent->maxs);
+			} else {
+				angles = vec3_origin;
+				UnpackBounds(ent->current.bounds, ent->mins, ent->maxs);
+			}
+
+			Cm_EntityBounds(ent->current.solid, ent->current.origin, angles, ent->mins, ent->maxs, ent->abs_mins, ent->abs_maxs);
+
+			Matrix4x4_CreateFromEntity(&ent->matrix, ent->current.origin, angles, 1.0);
+			Matrix4x4_Invert_Simple(&ent->inverse_matrix, &ent->matrix);
+		}
 	}
```

Comment:
Comment says it all, really. Changing `ent->current.var` to `ent->var` here yields pixel-perfect collision for shadows and whatnot, but this makes the prediction / error interpolation code go awry. I was seeing strange error correction when riding the elevator on Edge, particularly downward. So I've left it so that it works as the old code did, snapping entities' collision states against the server frame.

In thinking about how the prediction code works, I think I have some understanding of the problem. The server "thinks" in discrete 25ms snaps, while the client runs at e.g. 8ms snaps (125fps). Therefore, the client is going to predict against different snapshots of entity state than the server is. In theory, you'd think this would all even out in the end, but I guess it doesn't.